Add ServiceFabricDeploy task

mthalman-msft opened the pull request:
Adding two tasks to support Azure Service Fabric applications:

- ServiceFabricUpdateAppVersions: Utility task capable of automatically updating the versions in the manifest files of a Service Fabric application.
- ServiceFabricDeploy: Deploy task which deploys a Service Fabric application package to a cluster using AAD for authentication.

msftclas commented:
Hi __@mthalman-msft__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@mthalman-msft__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

ericsciple commented on Tasks/ServiceFabricDeploy/utilities.ps1:
```diff
@@ -0,0 +1,114 @@
+        
+        [String]
+        $Pattern
+    )
```
The comparisons here look fragile. I would recommend checking the comparisons differently, perhaps:
```powershell
if (@($Items).Length -gt 1) { [...] }
elseif ($Items -eq $null -or @($Items).Length -eq 0) { [...] }
```

IIRC PowerShell documentation is explicitly vague exactly what type is returned when capturing the result of a pipeline with multiple values. Whether it's System.Array, System.ArrayList, or some other similar type of collection is never really clear in PowerShell.

The way the books recommend to ensure the result is an array (using the term array loosely), is to wrap it in `@( )`. If it's already an array, then it's not wrapped again. If it's a single element, then it's wrapped in an array.

Part of the fragility here also is that you're taking a dependency on the task lib to return a System.Array from Find-VstsFiles.

ericsciple commented on Tasks/ServiceFabricDeploy/deploy.ps1:
```diff
@@ -0,0 +1,87 @@
+    $applicationPackagePathSearchPattern = Get-VstsInput -Name applicationPackagePath -Require
+    Write-Host (Get-VstsLocString -Key SearchingForApplicationPackage -ArgumentList $applicationPackagePathSearchPattern)
+
+    # Workaround for bug in Find-VstsFiles that doesn't handle the IncludeDirectories parameter correctly. Modifies the search pattern to look for app manifest file instead of the directory
```
I merged this fix a couple weeks ago. Although it doesn't look like the new version was published yet. I'll follow-up on that.

ericsciple replied:
If you rebase onto latest master, it's there now.

ericsciple commented on Tasks/ServiceFabricDeploy/utilities.ps1:
```diff
@@ -0,0 +1,114 @@
+        $PublishProfileFile
+    )
+
+    $publishProfileXml = [Xml] (Get-Content $PublishProfileFile)
```
Probably fine either way, but I always prefer LiteralPath.

ericsciple commented on Tasks/ServiceFabricDeploy/deploy.ps1:
```diff
@@ -0,0 +1,87 @@
+    }
+    else
+    {
+        foreach ($scriptFile in Get-ChildItem $moduleFolderPath -Filter "*.ps1")
```
Is there not a psm1/psd1 file to import?

mthalman-msft replied:
There is but it can't be used in this context.  It's due to the way in which PowerShell scopes its variables.  The functions contained in the modules need access to a global connection object variable which only works if you dot source the file versus importing.

ericsciple replied:
Variable inheritance within modules does inherit from the global scope. If you define the variable as global (`$global:foo = 'bar') then it will have access.

mthalman-msft replied:
These scripts don't own the variable so nothing can be done here.  The variable is set as a result of calling a cmdlet from the Service Fabric module.

ericsciple commented on Tasks/ServiceFabricDeploy/deploy.ps1:
```diff
@@ -0,0 +1,87 @@
+    $publishProfile = Read-PublishProfile $publishProfilePath
+    $clusterConnectionParameters = $publishProfile.ClusterConnectionParameters
+    
+    if (-not $clusterConnectionParameters.ContainsKey("AzureActiveDirectory") -or $clusterConnectionParameters.Item("AzureActiveDirectory") -eq $false)
```
In general, you might want to double check that the syntax you use for hashtables and xml works on PowerShell 3. All of our tasks support as far back as that version. There has been discussion of adding a minimum PowerShell version to the handler section of the task.json, but is not there today. Otherwise if there is a reason it can't be supported back to PowerShell 3, then the task should fail gracefully (e.g. check psversiontable and error out with a good message).

I can't remember if the syntax for accessing hashtable values using property syntax (i.e. `$hash.MyKey` over `$hash['MyKey']`) is supported in PS3. If not, then one option to enable the consumer to still use the property syntax is to use `New-Object -TypeName psobject -Property $hash` which will return an object will properties.

mthalman-msft replied:
PS3, and at least PS2, does support accessing key values using the property syntax.

ericsciple commented on Tasks/ServiceFabricUpdateAppVersions/version.ps1:
```diff
@@ -0,0 +1,74 @@
+[CmdletBinding()]
+param()
+
+function Assert-SingleItem
```
Same comment as the `Assert-SingleItem` in the other file.

ericsciple commented on Tasks/ServiceFabricDeploy/deploy.ps1:
```diff
@@ -0,0 +1,84 @@
+    $regKey = "HKLM:\SOFTWARE\Microsoft\Service Fabric SDK"
+    $moduleFolderPath = (Get-ItemProperty -Path $regKey -Name FabricSDKPSModulePath).FabricSDKPSModulePath
+
+    if (Test-Path "$moduleFolderPath\ServiceFabricSDK.ps1")
```
This logic really concerns me. You're finding *some version* of the module on disk and assuming internal details of the file structure of the module will hold true. I'm not familiar with this module, but this looks very fragile to me. How do you know that a psd1/psm1 won't contain critical logic in a future version of this module? Do you guys own this module (I'm not asking rhetorically, I am not familiar with this module and suppose you guys may actually own it)?

Have you considered bundling the module with your task? That way you can bundle a known version with the task where you know this logic works.

mthalman-msft replied:
This module is owned by the Service Fabric platform team and is installed on the build agents through their SDK.  It provides the actual deployment logic that we depend on.  We coordinate with them for any changes in this module so we're always aware and can react accordingly.

ericsciple replied:
That's not true for on-premises releases of TFS.

mthalman-msft replied:
How is that different than any other platform dependency?  It would require the on-premises system to have the SDK installed.

ericsciple replied:
discussed options offline

ericsciple commented on Tasks/ServiceFabricDeploy/deploy.ps1:
```diff
@@ -0,0 +1,84 @@
+    Write-Host (Get-VstsLocString -Key ConnectedToCluster)
+    
+    $regKey = "HKLM:\SOFTWARE\Microsoft\Service Fabric SDK"
+    $moduleFolderPath = (Get-ItemProperty -Path $regKey -Name FabricSDKPSModulePath).FabricSDKPSModulePath
```
It looks like the end user just ends up with a reg key not found error if this isn't installed. It would be better to fail with an actionable error message.

ericsciple commented on Tasks/ServiceFabricDeploy/utilities.ps1:
```diff
@@ -0,0 +1,123 @@
+function Read-PublishProfile
+{
+    Param (
+        [ValidateScript({Test-Path $_ -PathType Leaf})]
```
Does this syntax work in PS 3? I know before 5 (or 4?), creating a scriptblock required more cumbersome syntax (e.g. `[scriptblock]::create('string')` versus `{ some cmd }`. Although I'm not sure about within ValidateScript specifically.

mthalman-msft replied:
Removed that since the path is already being validated earlier on with a call to Assert-VstsPath.

chrisrpatterson commented on Tasks/ServiceFabricDeploy/task.loc.json:
```diff
@@ -38,20 +38,12 @@
-      "type": "string",
-      "label": "ms-resource:loc.input.label.aadUserName",
+      "name": "aadServiceConnectionName",
+      "type": "connectedService:Generic",
```
Needs to be regenerated after the change for the connected service type

mthalman-msft commented:
@ericsciple: Updated with another commit that addresses review comments.  This incorporates the Service Fabric SDK scripts that this task was previously dependent upon being installed on the build agent.  So now these scripts are embedded as part of the task implementation, making it more self-contained.

ericsciple commented on Tasks/ServiceFabricDeploy/ServiceFabricSDK/Utilities.ps1:
```diff
@@ -0,0 +1,207 @@
+    try 
+    {
+        [System.Reflection.Assembly]::LoadWithPartialName("System.IO.Compression.FileSystem") | Out-Null 
+        [System.IO.Compression.ZipFile]::ExtractToDirectory("$File", "$Destination") 
```
This won't work unless .Net 4.5.2 or higher is installed on the machine, correct?

ericsciple replied:
Consider try/catch if the DLL/type can't be loaded, and fail with an actionable error message.

ericsciple replied:
One of the other tasks, `ExtractFiles`, leverages 7zip. That might be a better option. @jasholl would have more details if you decide to go that route. He would know the license details.

ericsciple replied:
Actually, depending on .Net >=4.5.2 may be OK. Anyway, any task that consumes the VSTS webapi DLLs or extended client DLLs would require at least that high of a version.

mthalman-msft replied:
So, are you saying no changes are needed for the zip code?

ericsciple replied:
@bryanmacfarlane  @chrisrpatterson any thoughts here regarding guidance for task authors needing to unzip a file? I checked code in VSO, it looks like client SDKs target 4.5 only. IIRC the zip functionality didn't come to .Net until 4.5.2.

ericsciple replied:
I WAS WRONG AND @chrisrpatterson WAS RIGHT! It was introduced in .Net 4.5, not 4.5.2. So you are good here.

ericsciple commented on Tasks/ServiceFabricUpdateAppVersions/task.json:
```diff
@@ -2,7 +2,11 @@
   "id": "97EF6E59-B8CC-48AA-9937-1A01E35E7584",
   "name": "ServiceFabricUpdateAppVersions",
   "friendlyName": "Update Service Fabric App Versions",
+<<<<<<< b515e447160f0c9fd44418891c7e51eda1263d83
```
Oops

ericsciple commented on Tasks/ServiceFabricUpdateAppVersions/Strings/resources.resjson/en-US/resources.resjson:
```diff
@@ -5,6 +5,7 @@
   "loc.instanceNameFormat": "Update Service Fabric App Versions",
   "loc.input.label.applicationPackagePath": "Application Package",
   "loc.input.help.applicationPackagePath": "Path to the application package. [Variables](https://go.microsoft.com/fwlink/?LinkID=550988) and wildcards can be used in the path.",
+>>>>>>> Add ServiceFabricDeploy and ServiceFabricUpdateAppVersions tasks
```
Oops

ericsciple commented on Tasks/ServiceFabricDeploy/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1:
```diff
@@ -0,0 +1,271 @@
+﻿function Publish-NewServiceFabricApplication
```
I'm guessing these SDK files are largely copy/paste from a different repo?

Fyi - This reminds me, if you find yourself needing to share common code between different Service Fabric tasks, you can put it under `/Tasks/Common` and add an entry under `/common.json`. It will copy the common code into place at build time. For example, there is common code shared between the MSBuild task and VSBuild task.

mthalman-msft replied:
Yes, they're copy-pasted.  Thanks for the info on shared task code.  At this point, there should be no need to have this code shared with any other task.

ericsciple commented on Tasks/ServiceFabricDeploy/task.json:
```diff
@@ -0,0 +1,96 @@
+  "friendlyName": "Service Fabric Application Deployment",
+  "description": "Deploy a Service Fabric application to a cluster.",
+  "author": "Microsoft Corporation",
+  "helpMarkDown": "",
```
Our tasks typically have an fwlink to the documentation. It would be good for you to check with @chrisrpatterson if he has some recommendation for you here. I'm not sure what he has in mind.

mthalman-msft commented:
@ericsciple: Ok, another commit to review.  I also realized that the task should not have an azureps demand so I've removed that; it just needs plain old PowerShell.

ericsciple commented:
@mthalman-msft Cool. Is this one ready to merge? Would you mind squashing so history is useful when investigating issues?
